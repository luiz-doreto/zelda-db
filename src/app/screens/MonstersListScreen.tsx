import useMonsterViewModel from '@/viewModels/useMonstersVM';
import MonsterCard from '../components/MonsterCard';
import ZeldaList from '../components/ZeldaList';

const MonstersListScreen = () => {
  const { data, isLoading, error } = useMonsterViewModel();

  return (
    <ZeldaList
      data={data}
      isLoading={isLoading}
      error={error}
      renderItem={item => <MonsterCard monster={item} />}
    />
  );
};

export default MonstersListScreen;
