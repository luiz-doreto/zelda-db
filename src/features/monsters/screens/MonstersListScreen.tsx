import { StyleSheet, View } from 'react-native';

import MonsterCard from '../components/MonsterCard';
import Text from "@/components/Text";
import { FlashList } from '@shopify/flash-list';
import { SafeAreaView } from 'react-native-safe-area-context';
import useMonsterViewModel from '../viewModels/useMonstersViewModel';

const MonstersScreen = () => {

  const { data, isLoading, error } = useMonsterViewModel();

  if (isLoading) return <Text>Loading...</Text>;
  if (error) return <Text>Error: {error?.status} {JSON.stringify(error?.message)}</Text>;

  return (
    <SafeAreaView edges={['bottom', 'left', 'right']} style={{flex: 1}}>
        {
          data?.data.length === 0
            ? <Text>No Monsters</Text>
            : <FlashList
                contentContainerStyle={styles.listContentContainer}
                data={data?.data}
                renderItem={({ item }) => <MonsterCard key={item.id} monster={item}/>}
                ItemSeparatorComponent={() => <View style={styles.listSeparator}/>}
              />
        }
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  titleContainer: {
    alignItems: 'center',
    gap: 8,
    marginVertical: 20,
  },
  listContentContainer: {
    marginTop: 12,
  },
  listSeparator: {
    height: 12,
  },
});

export default MonstersScreen;
