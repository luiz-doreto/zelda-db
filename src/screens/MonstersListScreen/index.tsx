import MonsterCard from '~/components/MonsterCard';
import ZeldaList from '~/components/ZeldaList';
import useMonsterViewModel from './view.model';

const MonstersListScreen = () => {
  const { data, isLoading, error } = useMonsterViewModel();

  return (
    <ZeldaList
      data={data}
      isLoading={isLoading}
      error={error}
      renderItem={item => <MonsterCard monster={item} />}
    />
  );
};

export default MonstersListScreen;
