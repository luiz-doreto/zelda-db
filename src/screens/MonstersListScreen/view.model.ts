import { useEffect, useState } from 'react';
import { hooksApi, HooksApi } from '~/api/hooks';
import useHeaderSearch from '~/hooks/useHeaderSearch';
import { normalizeError } from '~/utils/rtkQueryErrorNormalizer';

const useMonsterViewModel = (hooks: HooksApi = hooksApi) => {
  const { data, isLoading, error } = hooks.useGetAllMonstersQuery();
  const { searchTerm } = useHeaderSearch();
  const [filteredData, setFilteredData] = useState(data?.data);

  useEffect(() => {
    if (searchTerm) {
      const filtered = data?.data?.filter(item =>
        item.name.toLowerCase().includes(searchTerm.toLowerCase())
      );
      setFilteredData(filtered);
    } else {
      setFilteredData(data?.data);
    }
  }, [searchTerm, data]);
  /**
   * Place all the logic with data.
   * connects with a secondary slice if needed
   */

  return {
    data: filteredData,
    isLoading,
    error: normalizeError(error),
  };
};

export default useMonsterViewModel;
