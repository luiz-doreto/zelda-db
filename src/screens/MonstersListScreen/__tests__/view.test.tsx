import { render } from '@testing-library/react-native';
import MonstersListScreen from '../index';
import useMonsterViewModel from '../view.model';

jest.mock('../view.model', () => jest.fn());
(useMonsterViewModel as jest.Mock).mockImplementation(() => ({
  data: [{ id: 1, name: 'myMonster' }],
  isLoading: false,
  error: null,
}));

describe('MonstersListScreen', () => {
  it('should match snapshot', () => {
    const { toJSON } = render(<MonstersListScreen />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should match snapshot when empty', () => {
    (useMonsterViewModel as jest.Mock).mockImplementation(() => ({
      data: [],
      isLoading: false,
      error: null,
    }));
    const { toJSON } = render(<MonstersListScreen />);
    expect(toJSON()).toMatchSnapshot();
  });
});
