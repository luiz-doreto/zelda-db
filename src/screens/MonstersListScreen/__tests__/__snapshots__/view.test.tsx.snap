// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonstersListScreen should match snapshot 1`] = `
<View>
  <Text>
    Monster:
    myMonster
  </Text>
</View>
`;

exports[`MonstersListScreen should match snapshot when empty 1`] = `
<Text>
  List Empty
</Text>
`;
